modin/experimental/core/storage_formats/omnisci/query_compiler.py

```diff
@@ -204,16 +204,19 @@ def from_arrow(cls, at, data_cls):
     # Dataframe exchange protocol
 
     def to_dataframe(self, nan_as_null: bool = False, allow_copy: bool = True):
-        raise NotImplementedError(
-            "The selected execution does not implement the DataFrame exchange protocol yet."
+        return self._modin_frame.__dataframe__(
+            nan_as_null=nan_as_null, allow_copy=allow_copy
         )
 
     @classmethod
     def from_dataframe(cls, df, data_cls):
-        raise NotImplementedError(
-            "The selected execution does not implement the DataFrame exchange protocol yet."
+        from modin.core.dataframe.base.exchange.dataframe_protocol.from_dataframe import (
```

Comment:
I'm not sure about this. Core dataframes implement the exchange protocol, which includes implementing `__dataframe__()` method only. We don't want to overcomplicate low-level frames with such routines as extracting the data from other protocol objects. 

Import implementation should be at higher levels I think, and in my opinion query compilers is a fine place for this.

Reply:
from_dataframe is apparently a part of the protocol. One of the reason to have this at the code dataframe layer is that we may want to interchange data between Modin core dataframe implementations using this protocol. See also this discussion ([link](https://github.com/modin-project/modin/pull/4150#discussion_r799512750)).

Reply:
Core dataframes already have import methods (`from_arrow`, `from_pandas`, [`from_partitions`](https://github.com/modin-project/modin/blob/006ccd0692169a3af38c44bd54216319ea6fccf6/modin/distributed/dataframe/pandas/partitions.py#L92) - this one is used for high-level frame, but I think it's doable to reuse the logic and build a low-level one), so my intention was to use QueryCompiler to build storage format specific objects from the exchange protocol (pandas DataFrame, pyarrow table, etc?...) and then use them as arguments for the existing core dataframe import methods.

For example, for Omnisci execution `qc.from_dataframe` builds zero-copy pyarrow.Table and the passes it to the `modin_frame.from_arrow`. For `PandasOn*` execution, `qc.from_dataframe` build a pandas DataFrame from each chunk, wraps these chunks into partitions, and then build a `modin_dataframe` from them.

Originally, it was intended to keep core dataframe as simple as possible to ease integration of new executions, IMO bringing `from_dataframe` method into the core dataframe's API unnecessarily complicates that, as it could be implemented in higher levels.

The main con of this approach that was mentioned in the discussion, is that we can't exchange the data between arbitrary dataframes. The answer is: we actually would never meet a case when we would need it. Usually, we don't operate with standalone core dataframe objects, we operate them in the context of an execution (qc + dataframe), so there's always a query compiler that's ready to import the data from an arbitrary exchange object into selected core dataframe.

Would also like to hear other opinions about this
@modin-project/modin-core

Reply:
Execution-specific dataframe classes are independent from each other. I think we should keep the protocol implementation exactly at this layer as it is the lowest layer for the protocol to be implemented.

> The answer is: we actually would never meet a case when we would need it.

That is not true. There might be two systems/applications, for example. The first one operates on a PandasOnRay dataframe, second one operates on a OmnisciOnNative dataframe. If those apps were to need to exchange data between each other, the dataframes would requires themselves to have the protocol implementations.

Reply:
Makes sense. Moved `from_dataframe` logic to a core dataframe layer in the last commit